great_expectations/data_context/data_context_variables.py

```diff
@@ -47,50 +53,56 @@ class DataContextVariables(ABC):
-    evaluation_parameter_store_name: Optional[str] = None
-    checkpoint_store_name: Optional[str] = None
-    profiler_store_name: Optional[str] = None
-    plugins_directory: Optional[str] = None
-    stores: Optional[dict] = None
-    data_docs_sites: Optional[dict] = None
-    notebooks: Optional[NotebookConfig] = None
-    config_variables_file_path: Optional[str] = None
-    anonymous_usage_statistics: Optional[AnonymizedUsageStatisticsConfig] = None
-    concurrency: Optional[ConcurrencyConfig] = None
-    progress_bars: Optional[ProgressBarsConfig] = None
-    _store: Optional["DataContextVariablesStore"] = None  # noqa: F821
+    config: DataContextConfig
+    substitutions: Optional[dict] = None
+    _store: Optional["DataContextStore"] = None  # noqa: F821
+
+    def __post_init__(self) -> None:
+        if self.substitutions is None:
+            self.substitutions = {}
 
     @property
-    def store(self) -> "DataContextVariablesStore":  # noqa: F821
+    def store(self) -> "DataContextStore":  # noqa: F821
         if self._store is None:
             self._store = self._init_store()
         return self._store
 
     @abstractmethod
-    def _init_store(self) -> "DataContextVariablesStore":  # noqa: F821
+    def _init_store(self) -> "DataContextStore":  # noqa: F821
         raise NotImplementedError
 
-    def _get_key(self, attr: DataContextVariableSchema) -> DataContextVariableKey:
-        # Chetan - 20220607 - As it stands, DataContextVariables can only perform CRUD on entire objects.
-        # The DataContextVariablesKey, when used with an appropriate Store and StoreBackend, gives the ability to modify
-        # individual elements of nested config objects.
-
-        key: DataContextVariableKey = DataContextVariableKey(resource_type=attr.value)
+    def get_key(self) -> DataContextKey:
+        """
+        Generates the appropriate Store key to retrieve/store configs.
+        """
+        key: ConfigurationIdentifier = ConfigurationIdentifier(
+            configuration_key=DataContextVariableSchema.DATA_CONTEXT
+        )
         return key
 
     def _set(self, attr: DataContextVariableSchema, value: Any) -> None:
-        setattr(self, attr.value, value)
-        key: DataContextVariableKey = self._get_key(attr)
-        self.store.set(key=key, value=value)
+        key: str = attr.value
+        self.config[key] = value
 
     def _get(self, attr: DataContextVariableSchema) -> Any:
-        val: Any = getattr(self, attr.value)
-        return val
+        key: str = attr.value
+        val: Any = self.config[key]
+        substituted_val: Any = substitute_config_variable(val, self.substitutions)
+        return substituted_val
+
+    def save_config(self) -> None:
+        """
+        Persist any changes made to variables utilizing the configured Store.
+        """
+        key: ConfigurationIdentifier = self.get_key()
+        self.store.set(key=key, value=self.config)
```

Comment:
We previously persisted values with each `_set` call - this is not desired behavior.

The user should be able to make modifications as they see fit and ONLY when they deem it appropriate, explicitly invoke the `save_config` command to persist changes using the store. This is a subtle change but is more in-line with how GE currently works.